### ISSUE TITLE
Bug: Guile 2.0.13 not compatible with meep 1.3

### ISSUE THREAD

**Issue by JohnWeiner** [stevengj/meep#57]
I just installed meep and related prerequisites via Filip Dominic's configure script, python-meep-install. The first tutorial (1D waveguide), which I ran as tutorial1.ctl, reports errors such as:

Using MPI version 3.1, 1 processes
-----------
Initializing structure...
Working in 2D dimensions.
Computational cell is 16 x 8 x 0 with resolution 10
     block, center = (0,0,0)
          size (1e+20,1,1e+20)
          axes (1,0,0), (0,1,0), (0,0,1)
          dielectric constant epsilon diagonal = (12,12,12)
time for set_epsilon = 0.163085 s
-----------
creating output file "./tutorial1-eps-000000.00.h5"...
Backtrace:
In ice-9/boot-9.scm:
 160: 12 [catch #t #<catch-closure 55a6ef700260> ...]
In unknown file:
   ?: 11 [apply-smob/1 #<catch-closure 55a6ef700260>]
In ice-9/eval.scm:
 432: 10 [eval # #]
 432: 9 [eval # #]
In unknown file:
   ?: 8 [primitive-load "tutorial1.ctl"]
In ice-9/eval.scm:
 432: 7 [eval # #]
In ice-9/boot-9.scm:
 703: 6 [map #<procedure 55a6efc04b40 at ice-9/eval.scm:416:20 (a)> #]
In ice-9/eval.scm:
 399: 5 [eval # #]
 387: 4 Exception thrown while printing backtrace:
ERROR: In procedure delete-meep-volume: Wrong type argument in position 1: #<finalized smob 55a6f00f4a00>

ERROR: In procedure %run-finalizers:
ERROR: In procedure delete-meep-volume: Wrong type argument in position 1: #<finalized smob 55a6f00f4de0>

Some deprecated features have been used.  Set the environment
variable GUILE_WARN_DEPRECATED to "detailed" and rerun the
program to get more information.  Set it to "no" to suppress
this message.

Filip Dominic found that everything ran without error on his local computer using Guile 2.0.11 but he reproduced the errors when he ran the same ctl file using Guile 2.0.13.  He asked me to report this bug in 2.0.13.

**Comment by oskooi** [comment]
The issue with Meep's incompatibility with the latest version of Guile has to do with SWIG's outdated support for Guile. See this [SWIG issue](https://github.com/swig/swig/issues/312) for more details. Unfortunately, the only workaround right now seems to be using older versions of Guile.

**Comment by oskooi** [comment]
Closed #57.

**Comment by JohnWeiner** [comment]
It appears from a lot of cut-and-try with a lot of different versions, that the latest version one can use is guile-2.0.11.  That one works with meep on macOS and on Ubuntu/linux.

John

**Comment by JohnWeiner** [comment]
I think it would be a good idea to edit the web pages on the meep Ab Initio web site that warmly recommends to Debian Ubuntu/linux users to download the binary package using apt-get install.   Because of the guile version incompatibility, new users will quickly come to grief and give up in discouragement.   Until the SWIG problem gets fixed, downloading and building the requisite tar balls or cloning from the GitHub might be the only viable way to get a functioning meep on Debian/Ubuntu or macOS.

**Comment by JohnWeiner** [comment]
Hi Daniel,

The version Guile 2.0.13 does not seem to be compatible with meep 1.3.  The symptom is the error,

> ERROR: In procedure delete-meep-volume: Wrong type argument in position 1: #<finalized smob 55d264518180>
I have been able to get things working with Guile 2.0.11 but no version more recent than that.  If you can tell the Debian package handler to use Guile 2.0.11 then I would try that.  If not, you can install from sources, but that is a somewhat more arduous process.

Best regards,

John Weiner

**Comment by Daniel-M** [comment]
Thanks John, I've tried to build from sources but the process fails too.
Look at https://github.com/stevengj/meep/issues/94.

I'll try to downgrade the guile version and try again.


-- 

*Daniel Mejía R.*

*Las más formidables armas del hombre para su conquista del Conocimiento
son la mente racional y la insaciable curiosidad que lo impulsa...*
Isaac Asimov


​
<http://www.udea.edu.co/wps/portal/udea/web/inicio/investigacion/grupos-investigacion/ciencias-naturales-exactas/biofisica>
<http://www.udea.edu.co/wps/portal/udea/web/inicio/investigacion/grupos-investigacion/ciencias-naturales-exactas/biofisica>Grupo
de Biofísic​a - Universidad de Antioquia
<http://www.udea.edu.co/wps/portal/udea/web/inicio/investigacion/grupos-investigacion/ciencias-naturales-exactas/biofisica>

​
​
http://danielmejia.me​

​ResearchGate: Daniel_Mejia_Raigosa​
<https://www.researchgate.net/profile/Daniel_Mejia_Raigosa>

​ <http://danielmejia.me>*bitbucket* <http://github.com/Daniel-M>*:*
Daniel-M <http://github.com/Daniel-M>
*github:* Daniel-M <http://github.com/Daniel-M>
*gitlab:* Daniel-M <http://github.com/Daniel-M>
​--

2017-09-07 9:46 GMT-05:00 John Weiner <notifications@github.com>:

**Comment by Daniel-M** [comment]
***UPDATE*** Manually installed Guile v2.0.11, created an alternative for guile with `update-alternatives` in order to provide a way to change between guile version on my system without breaking anything. Once guile v2.0.11 was picked as the system wide guile, meep kept failing with the same error initially reported.

**Comment by Daniel-M** [comment]
Purged any non v2.0.11 version of guile and tried again with no sucess. I can't use meep nor compiled since the problems detailed at #94 won't allow me to do so.

**Comment by JohnWeiner** [comment]
Hi Daniel

The errors don’t look like Guile problems.  Make sure that ALL the prerequisites (libctl, h5utils, etc)  are installed cleanly, using the Github “git clone” followed by invoking the autogen.sh script to generate the configure file, when possible.  The “how to” is explained in the “README” and “INSTALL” files.  Some of the prerequisites might have to be installed from tar balls.  Keep the guile 2.0.11, you will need it.    On macOS Sierra, a number of meep users have found that that they can get meep 1.3 itself installed from the tarball on the Ab Inito site, but that the installation from the GitHub repository fails (but not with kind of error messages you are getting).

Best regards,  John

**Comment by Daniel-M** [comment]
Hi @JohnWeiner,

What I have done so far is,

**Prerequisites**

* Installed libctl from the master (and only) branch of https://github.com/stevengj/libctl.  
* Installed guile v2.0.11 downloaded from https://ftp.gnu.org/gnu/guile/guile-2.0.11.tar.gz  
* Installed meep v1.3 from master branch of this repo   (as refered in the closed #94 )
* Installed from Debian official repository: libblas-dev liblapack-dev libunistring-dev h5utils libpng-dev 
* GSL and GMP manually installed to their latest versions. (libctl requires GMP, It wouldn't have installed other wise)

So far those are the recommendations by readthedocs I recall.

**MEEP behavior**

The output of `meep --version` is,

```sh
Meep 1.3, Copyright (C) 2005-2015 Massachusetts Insitute of Technology.
Using libctl 3.3.0 and Guile 2.0.11.
```

Tried to run the following script for a waveguide example,

```sh
#Contents of waveguide.ctl 
(reset-meep)
(set! geometry-lattice (make lattice (size 16 8 no-size)))
(set! geometry (list
                (make block (center 0 0) (size infinity 1 infinity)
                      (material (make dielectric (epsilon 12))))))
(set! sources (list
               (make source
                 (src (make continuous-src (frequency 0.15)))
                 (component Ez)
                 (center -7 0))))
(set! pml-layers (list (make pml (thickness 1.0))))
(set! resolution 10)
(run-until 200
           (at-beginning output-epsilon)
           (at-end output-efield-z))
```

The output was,

```sh
$ meep waveguide.ctl 
-----------
Initializing structure...
Working in 2D dimensions.
Computational cell is 16 x 8 x 0 with resolution 10
     block, center = (0,0,0)
          size (1e+20,1,1e+20)
          axes (1,0,0), (0,1,0), (0,0,1)
          dielectric constant epsilon diagonal = (12,12,12)
time for set_epsilon = 0.0403681 s
-----------
creating output file "./waveguide-eps-000000.00.h5"...
creating output file "./waveguide-ez-000200.00.h5"...
run 0 finished at t = 200.0 (4000 timesteps)

Elapsed run time = 1.62755 s

Some deprecated features have been used.  Set the environment
variable GUILE_WARN_DEPRECATED to "detailed" and rerun the
program to get more information.  Set it to "no" to suppress
this message.
```

What am I doing wrong?

**Comment by JohnWeiner** [comment]
I don’t think you are doing anything wrong.  The waveguide script terminated normally.  The “deprecated” message is not an error.  I set the environment variable as suggested in the message (export GUILE_WARN_DEPRECATED=no) in my .bash_profile so as not to be annoyed by this recurring “warning”.  Deprecated code is perfectly usable…the message just says that, in the opinion of the developer/maintainer, there is a better way to do things.  Check your h5utils with, e.g. h5topng on the h5 files that the waveguide code generated.

**Comment by stevengj** [comment]
With the latest master, I can compile Meep with the libctl front-end on MacOS with Guile 2.2.2 and it seems to be basically working, although there are a couple of odd warnings that I'm still tracking down.

**Comment by JohnWeiner** [comment]
Does that mean that SWIG has been updated to take account of changes in Guile since 2.0.11?

**Comment by ppfromero** [comment]
Hello I have some similar issue with meep in Ubuntu stable version over VirtualBox.
I make very simple file test.ctl from tutorials.

(reset-meep)
(set! geometry-lattice (make lattice (size 16 16 no-size)))
(set! geometry (list
                (make block (center -2 -3.5) (size 12 1 infinity)
                      (material (make dielectric (epsilon 12))))
                (make block (center 3.5 2) (size 1 12 infinity)
                      (material (make dielectric (epsilon 12))))))
(set! pml-layers (list (make pml (thickness 1.0))))
(set! resolution 10)
(set! sources (list
               (make source
                 (src (make continuous-src
                        (wavelength (* 2 (sqrt 11))) (width 20)))
                 (component Ez)
                 (center -7 -3.5) (size 0 1))))
(run-until 200
           (at-beginning output-epsilon)
           (to-appended "ez" (at-every 0.6 output-efield-z)))



And get the following out file:


-----------
Initializing structure...
Working in 2D dimensions.
Computational cell is 16 x 16 x 0 with resolution 10
     block, center = (-2,-3.5,0)
          size (12,1,1e+20)
          axes (1,0,0), (0,1,0), (0,0,1)
          dielectric constant epsilon diagonal = (12,12,12)
     block, center = (3.5,2,0)
          size (1,12,1e+20)
          axes (1,0,0), (0,1,0), (0,0,1)
          dielectric constant epsilon diagonal = (12,12,12)
time for set_epsilon = 0.038491 s
-----------
creating output file "./test-ez.h5"...
creating output file "./test-eps-000000.00.h5"...
Backtrace:
In ice-9/boot-9.scm:
 160: 12 [catch #t #<catch-closure 559cffcce280> ...]
In unknown file:
   ?: 11 [apply-smob/1 #<catch-closure 559cffcce280>]
In ice-9/eval.scm:
 432: 10 [eval # #]
 432: 9 [eval # #]
In unknown file:
   ?: 8 [primitive-load "test.ctl"]
In ice-9/eval.scm:
 432: 7 [eval # #]
In ice-9/boot-9.scm:
 703: 6 [map #<procedure 559d0085a750 at ice-9/eval.scm:416:20 (a)> #]
In ice-9/eval.scm:
 399: 5 [eval # #]
 387: 4 Exception thrown while printing backtrace:
ERROR: In procedure delete-meep-volume: Wrong type argument in position 1: #<finalized smob 559d008678a0>

ERROR: In procedure %run-finalizers:
ERROR: In procedure delete-meep-volume: Wrong type argument in position 1: #<finalized smob 559d007e70a0>

Some deprecated features have been used.  Set the environment
variable GUILE_WARN_DEPRECATED to "detailed" and rerun the
program to get more information.  Set it to "no" to suppress
this message.


I already post in diffusion list but any answer.... Do you have any idea about this kind of problem?

Thanks in advance

Bes regards

**Comment by ChristopherHogan** [comment]
@ppfromero, the Ubuntu packages are out of date, but we're working on updating them. In the meantime, you can try building the latest 1.5 release from source. Instructions are [here](http://meep.readthedocs.io/en/latest/Build_From_Source/#building-from-source).

**Comment by ppfromero** [comment]
Dear Christopher,

These instruction are made for python interface, which are the git repository for scheme?

Best regards


JOSE ROMERO  | Optoélectronique - Photonique - Traitement du signal - Image

Tel.: 06.66.38.49.87
1752 Chemin du pas de la mue
13170 Les Pennes Mirabeau
email: ppfromero@hotmail.com

________________________________
De : Christopher Hogan <notifications@github.com>
Envoyé : mercredi 11 juillet 2018 13:56
À : stevengj/meep
Cc : ppfromero; Mention
Objet : Re: [stevengj/meep] Bug: Guile 2.0.13 not compatible with meep 1.3 (#57)


@ppfromero<https://github.com/ppfromero>, the Ubuntu packages are out of date, but we're working on updating them. In the meantime, you can try building the latest 1.5 release from source. Instructions are here<http://meep.readthedocs.io/en/latest/Build_From_Source/#building-from-source>.

**Comment by ppfromero** [comment]
Dear Christopher and friends,

I follow instructions but in my system ubuntu libpng16 is not more available and replaced by libpng-dev:i386
and libpython3.5-dev is not available...

Regards

JOSE ROMERO  | Optoélectronique - Photonique - Traitement du signal - Image

Tel.: 06.66.38.49.87
1752 Chemin du pas de la mue
13170 Les Pennes Mirabeau
email: ppfromero@hotmail.com

________________________________
De : Steven G. Johnson <notifications@github.com>
Envoyé : mardi 17 juillet 2018 11:20
À : stevengj/meep
Cc : ppfromero; Mention
Objet : Re: [stevengj/meep] Bug: Guile 2.0.13 not compatible with meep 1.3 (#57)


Closed #57<https://github.com/stevengj/meep/issues/57> via #419<https://github.com/stevengj/meep/pull/419>.

**Comment by ChristopherHogan** [comment]
@ppfromero, if you're on ubuntu 17.10 or 18.04, then `libpng-dev` should work. You don't need to install any of the packages with `python` in the name, since you're only interested in the scheme interface.  I'll update the build instructions now that LTS 18.04 is out.
